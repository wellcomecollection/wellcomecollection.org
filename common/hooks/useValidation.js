import { useState } from 'react';

const useValidation = (initIsValid = false, initShowValidity = false) => {
  const [isValid, setIsValid] = useState(initIsValid);
  const [showValidity, setShowValidity] = useState(initShowValidity);

  return { isValid, setIsValid, showValidity, setShowValidity };
};

export default useValidation;
