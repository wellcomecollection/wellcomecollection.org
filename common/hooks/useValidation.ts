import { Dispatch, SetStateAction, useState } from 'react';

type Validation = {
  isValid: boolean;
  setIsValid: Dispatch<SetStateAction<boolean>>;
  showValidity: boolean;
  setShowValidity: Dispatch<SetStateAction<boolean>>;
};

const useValidation = (
  initIsValid = false,
  initShowValidity = false
): Validation => {
  const [isValid, setIsValid] = useState(initIsValid);
  const [showValidity, setShowValidity] = useState(initShowValidity);

  return { isValid, setIsValid, showValidity, setShowValidity };
};

export default useValidation;
