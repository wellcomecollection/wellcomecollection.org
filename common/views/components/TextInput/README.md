## Purpose
To display a text input and any relevant validation UI/messaging to the user.

### Notes
- The label minifies if input has focus
- The label remains minified when input is blurred as long as some value is entered
- The label is always minified if user doesn't have JS
- The input will validate when it is blurred or the form it lives in is submitted (either shows error state or 'valid' checkmark)
- The input will hide any error at the point when it becomes valid and won't re-check for errors until it is blurred
- The input will hide valid checkmark if it receives new input and won't revalidate until it is blurred

[Edit this on GitHub](https://github.com/wellcomecollection/wellcomecollection.org/edit/master/common/views/components/TextInput/README.md)

[View this on Zeplin](https://zpl.io/VQDJKEn)
