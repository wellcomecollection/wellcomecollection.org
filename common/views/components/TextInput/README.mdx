## Purpose
To display a text input and any relevant validation UI/messaging to the user.

### Notes
- The input will validate when it is blurred or the form it lives in is submitted (either shows error or success state)
- The input will hide any error at the point when it becomes valid and won't re-check for errors until it is blurred
- The input will hide valid checkmark if it receives new input and won't revalidate until it is blurred

[Edit this Readme on GitHub](https://github.com/wellcomecollection/wellcomecollection.org/edit/main/common/views/components/TextInput/README.mdx)
