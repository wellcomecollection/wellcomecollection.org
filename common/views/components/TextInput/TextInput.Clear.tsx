import { Dispatch, FunctionComponent, RefObject, SetStateAction } from 'react';
import styled from 'styled-components';

import { clear } from '@weco/common/icons';
import Icon from '@weco/common/views/components/Icon/Icon';

const Button = styled.button`
  position: absolute;
  line-height: 1;
  top: 50%;
  transform: translateY(-50%);
  padding: 0;
`;

type Props = {
  inputRef: RefObject<HTMLInputElement>;
  setValue: Dispatch<SetStateAction<string>>;
  clickHandler?: () => void;
  right?: number;
};

const ClearSearch: FunctionComponent<Props> = ({
  inputRef,
  setValue,
  right,
  clickHandler,
}: Props) => {
  return (
    <Button
      style={right ? { right: `${right}px` } : undefined}
      onClick={() => {
        setValue('');
        clickHandler && clickHandler();
        inputRef?.current?.focus();
      }}
      type="button"
      aria-label="clear"
    >
      <Icon iconColor="black" icon={clear} title="Clear" />
    </Button>
  );
};

export default ClearSearch;
