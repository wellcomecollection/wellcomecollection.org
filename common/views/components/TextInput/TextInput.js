// @flow
import styled from 'styled-components';
import { font } from '../../../utils/classnames';

const VisuallyHidden = styled.div`
  border: 0;
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
  white-space: nowrap;
`;

const StyledInput = styled.input`
  width: 100%;
  padding: 0.4em;
  border: 1px solid ${props => props.theme.colors.pumice};

  &:focus {
    outline: 0;
    border: 1px solid ${props => props.theme.colors.turquoise};
  }

  &::-ms-clear {
    display: none;
  }
`;

type Props = {
  label: string,
  // We can also pass inputProps here
};

// $FlowFixMe (forwardRef)
const TextInput = React.forwardRef(({ label, ...inputProps }: Props, ref) => ( // eslint-disable-line
  <label className="flex flex--v-center">
    <StyledInput
      ref={ref}
      className={font({ s: 'HNL3', m: 'HNL2' })}
      type="text"
      {...inputProps}
    />
    <VisuallyHidden>
      <label>{label}</label>
    </VisuallyHidden>
  </label>
));

TextInput.displayName = 'TextInput';

export default TextInput;
