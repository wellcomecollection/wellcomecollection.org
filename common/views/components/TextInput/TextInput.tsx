import { forwardRef, useContext } from 'react';
import styled from 'styled-components';
// $FlowFixMe (tsx)
import Icon from '../Icon/Icon';
import { AppContext } from '../AppContext/AppContext';
import { classNames } from '../../../utils/classnames';

type TextInputWrapProps = {
  value: string;
  big: boolean;
  hasErrorBorder: boolean;
};
const TextInputWrap = styled.div.attrs({
  className: classNames({
    'flex relative': true,
  }),
})<TextInputWrapProps>`
  font-size: ${props => (props.big ? '20px' : '16px')};

  &:focus-within {
    box-shadow: ${props => props.theme.focusBoxShadow};

    label {
      font-size: 14px;
      transform: translateY(0%);
      top: 4px;
    }
  }
  overflow: hidden;
  border-radius: 6px;

  ${props =>
    props.hasErrorBorder &&
    `
    box-shadow: 0 0 0 1px ${props.theme.color('red')};
  `}
`;

type TextInputLabelProps = {
  isEnhanced: boolean;
  hasValue: boolean;
};
const TextInputLabel = styled.label.attrs({
  className: classNames({
    absolute: true,
  }),
})<TextInputLabelProps>`
  /* Styles for browsers that don't support :focus-within (<=IE11) */
  font-size: 14px;
  transform: translateY(0%);
  top: 4px;

  /* IE doesn't support :focus-within, but you can't test for :focus-within
  using @supports. Fortunately, IE doesn't support @supports, so this only
  targets browsers that support @suports (> IE11) */
  @supports (display: block) {
    font-size: inherit;
    transform: translateY(-50%);
    top: 50%;
  }

  white-space: nowrap;
  left: 15px;
  transition: top 125ms ease-in, font-size 125ms ease-in,
    transform 125ms ease-in;
  pointer-events: none;
  color: ${props => props.theme.color('pewter')};

  ${props =>
    (!props.isEnhanced || props.hasValue) &&
    `
    @supports (display: block) {
      top: 4px;
      transform: translateY(0);
      font-size: 14px;
    }
  `}
`;

type TextInputInputProps = {
  hasErrorBorder: boolean;
  big: boolean;
};
const TextInputInput = styled.input.attrs(props => ({
  type: props.type || 'text',
}))<TextInputInputProps>`
  padding: ${props =>
    props.big ? '27px 130px 8px 15px' : '27px 40px 8px 15px'};
  appearance: none;
  border: 0;
  height: 100%;
  border: 1px solid ${props => props.theme.color('pumice')};
  border-radius: 6px;
  font-size: inherit;
  width: 100%;

  &:focus {
    outline: 0;
    border-color: ${props => props.theme.color('turquoise')};
  }

  &:-ms-clear {
    display: none;
  }

  ${props =>
    props.hasErrorBorder &&
    `
      &,
      &:focus {
        border-color: ${props.theme.color('red')};
      }
    `}
`;

const TextInputCheckmark = styled.span.attrs({
  'data-test-id': 'TextInputCheckmark',
  className: classNames({
    absolute: true,
  }),
})`
  top: 50%;
  transform: translateY(-50%);
  right: 10px;
  line-height: 0;
`;

const TextInputErrorMessage = styled.span.attrs({
  'data-test-id': 'TextInputErrorMessage',
  className: classNames({
    'font-hnb': true,
  }),
})`
  font-size: 14px;
  margin-top: 10px;
  padding-left: 15px;
  color: ${props => props.theme.color('red')};
`;

type Props = {
  label: string;
  value: string;
  setValue: (value: string) => void;
  id: string;
  name?: string;
  type?: string;
  pattern?: string;
  required?: boolean;
  placeholder?: string;
  errorMessage?: string;
  isValid?: boolean;
  setIsValid?: (value: boolean) => void;
  showValidity?: boolean;
  setShowValidity?: (value: boolean) => void;
  autoFocus?: boolean;
  big?: boolean;
  ariaLabel?: string;
};

const TextInput = forwardRef(
  (
    {
      label,
      type,
      value,
      setValue,
      id,
      name,
      pattern,
      required,
      errorMessage,
      isValid,
      setIsValid,
      showValidity,
      setShowValidity,
      autoFocus,
      big,
      ariaLabel,
    }: Props,
    ref: any
  ) => {
    const { isEnhanced } = useContext(AppContext);

    function onChange(event) {
      const isValueValid = event.currentTarget.validity.valid;

      setValue(event.currentTarget.value);

      if (isValid && setShowValidity) {
        setShowValidity(false);
      }

      setIsValid && setIsValid(isValueValid);

      if (isValueValid && setShowValidity) {
        setShowValidity(false);
      }
    }

    function onBlur(event) {
      setIsValid && setIsValid(event.currentTarget.validity.valid);
      setShowValidity && setShowValidity(!!value && true);
    }

    return (
      <div>
        <TextInputWrap
          value={value}
          hasErrorBorder={!!(!isValid && showValidity)}
          big={!!big}
        >
          <TextInputLabel
            isEnhanced={isEnhanced}
            hasValue={!!value}
            htmlFor={id}
          >
            {label}
          </TextInputLabel>
          <TextInputInput
            ref={ref}
            id={id}
            name={name}
            required={required}
            value={value}
            pattern={pattern}
            onChange={onChange}
            onBlur={onBlur}
            hasErrorBorder={!!(!isValid && showValidity)}
            type={type}
            autoFocus={autoFocus}
            aria-label={ariaLabel}
            big={!!big}
          />
          {isValid && showValidity && (
            <TextInputCheckmark>
              <Icon name={`check`} color={'green'} />
            </TextInputCheckmark>
          )}
        </TextInputWrap>
        <div role="status">
          {errorMessage && !isValid && showValidity && (
            <TextInputErrorMessage>{errorMessage}</TextInputErrorMessage>
          )}
        </div>
      </div>
    );
  }
);

TextInput.displayName = 'TextInput';

export default TextInput;
