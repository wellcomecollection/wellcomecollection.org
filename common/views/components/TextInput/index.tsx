import {
  forwardRef,
  RefObject,
  ForwardRefRenderFunction,
  ChangeEvent,
  FocusEvent,
} from 'react';
import styled from 'styled-components';
import Icon from '@weco/common/views/components/Icon/Icon';
import { check } from '@weco/common/icons';
import { font } from '@weco/common/utils/classnames';

type TextInputWrapProps = {
  $big?: boolean;
  $hasErrorBorder: boolean;
  $darkBg?: boolean;
};
export const TextInputWrap = styled.div.attrs<TextInputWrapProps>(props => ({
  className: props.$big ? font('intr', 4) : font('intr', 5),
}))<TextInputWrapProps>`
  display: flex;
  position: relative;
  border: ${props =>
    props.$hasErrorBorder
      ? `3px solid ${props.theme.color('validation.red')}`
      : `2px solid ${props.theme.color(
          props.$darkBg ? 'white' : 'neutral.600'
        )}`};

  &:focus-within {
    box-shadow: ${props => props.theme.focusBoxShadow};
  }

  overflow: hidden;
`;

type TextInputLabelProps = {
  $isEnhanced: boolean;
  $hasValue: boolean;
};
export const TextInputLabel = styled.label<TextInputLabelProps>`
  position: absolute;
  left: 15px;

  /* Styles for browsers that don't support :focus-within (<=IE11) */
  font-size: 14px;
  transform: translateY(0%);
  top: 4px;

  /* IE doesn't support :focus-within, but you can't test for :focus-within
  using @supports. Fortunately, IE doesn't support @supports, so this only
  targets browsers that support @supports (> IE11) */
  @supports (display: block) {
    font-size: inherit;
    transform: translateY(-50%);
    top: 50%;
  }

  color: ${props => props.theme.color('neutral.600')};
  white-space: nowrap;
  transition:
    top ${props => props.theme.transitionProperties},
    font-size ${props => props.theme.transitionProperties},
    transform ${props => props.theme.transitionProperties};
  pointer-events: none;

  ${props =>
    (!props.$isEnhanced || props.$hasValue) &&
    `
    @supports (display: block) {
      top: 4px;
      transform: translateY(0);
      font-size: 14px;
    }
  `}
`;

type TextInputInputProps = {
  $hasErrorBorder: boolean;
};
export const TextInputInput = styled.input.attrs<{ $type?: string }>(props => ({
  type: props.$type || 'text',
}))<TextInputInputProps>`
  padding: 17px 35px 17px 15px;
  appearance: none;
  border: 0;
  height: 100%;
  font-size: inherit;
  width: 100%;

  &::-ms-clear {
    display: none;
  }
`;

const TextInputCheckmark = styled.span.attrs<{
  'data-testid'?: 'TextInputCheckmark';
}>({
  'data-testid': 'TextInputCheckmark',
})`
  position: absolute;
  top: 50%;
  transform: translateY(-50%);
  right: 10px;
  line-height: 0;
`;

export const TextInputErrorMessage = styled.span.attrs({
  role: 'alert',
  'data-testid': 'TextInputErrorMessage',
  className: 'font-intb',
})`
  display: block;
  font-size: 14px;
  margin-top: 10px;
  padding-left: 15px;
  color: ${props => props.theme.color('validation.red')};
`;

type Props = {
  label: string;
  value: string;
  setValue: (value: string) => void;
  id: string;
  name?: string;
  type?: string;
  pattern?: string;
  required?: boolean;
  placeholder?: string;
  errorMessage?: string;
  isValid?: boolean;
  setIsValid?: (value: boolean) => void;
  showValidity?: boolean;
  setShowValidity?: (value: boolean) => void;
  autoFocus?: boolean;
  big?: boolean;
  ariaLabel?: string;
  ariaDescribedBy?: string;
  form?: string;
  darkBg?: boolean;
};

const Input: ForwardRefRenderFunction<HTMLInputElement, Props> = (
  {
    label,
    type,
    value,
    setValue,
    id,
    name,
    pattern,
    required,
    placeholder,
    errorMessage,
    isValid,
    setIsValid,
    showValidity,
    setShowValidity,
    autoFocus,
    ariaLabel,
    ariaDescribedBy,
    form,
    big,
    darkBg,
  }: Props,
  ref: RefObject<HTMLInputElement>
) => {
  function onChange(event: ChangeEvent<HTMLInputElement>) {
    const isValueValid = event.currentTarget.validity.valid;

    setValue(event.currentTarget.value);

    if (isValid && setShowValidity) {
      setShowValidity(false);
    }

    setIsValid && setIsValid(isValueValid);

    if (isValueValid && setShowValidity) {
      setShowValidity(false);
    }
  }

  function onBlur(event: FocusEvent<HTMLInputElement>) {
    setIsValid && setIsValid(event.currentTarget.validity.valid);
    setShowValidity && setShowValidity(!!value && true);
  }

  return (
    <div>
      <TextInputWrap
        $hasErrorBorder={!!(!isValid && showValidity)}
        $big={!!big}
        $darkBg={darkBg}
      >
        <label className="visually-hidden" htmlFor={id}>
          {label}
        </label>
        <TextInputInput
          ref={ref}
          id={id}
          name={name}
          required={required}
          value={value}
          pattern={pattern}
          onChange={onChange}
          onBlur={onBlur}
          placeholder={placeholder || label}
          autoFocus={autoFocus}
          $type={type}
          $hasErrorBorder={!!(!isValid && showValidity)}
          aria-label={ariaLabel}
          aria-describedby={ariaDescribedBy}
          aria-invalid={!!(!isValid && showValidity)}
          aria-errormessage={errorMessage && `${id}-errormessage`}
          form={form}
        />
        {isValid && showValidity && (
          <TextInputCheckmark>
            <Icon icon={check} iconColor="validation.green" />
          </TextInputCheckmark>
        )}
      </TextInputWrap>
      {errorMessage && !isValid && showValidity && (
        <TextInputErrorMessage id={`${id}-errormessage`} role="alert">
          {errorMessage}
        </TextInputErrorMessage>
      )}
    </div>
  );
};

const TextInput = forwardRef<HTMLInputElement, Props>(Input);

export default TextInput;
