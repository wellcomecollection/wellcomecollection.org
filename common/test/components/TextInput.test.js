import { useState } from 'react';
import TextInput from '../../views/components/TextInput/TextInput';
import { mountWithTheme } from '../fixtures/enzyme-helpers';
import useValidation from '../../hooks/useValidation';

const ExampleTextInput = () => {
  const [value, setValue] = useState('');

  return (
    <TextInput
      type="email"
      value={value}
      setValue={setValue}
      required={true}
      errorMessage={`test error message`}
      {...useValidation()}
    />
  );
};

describe('TextInput', () => {
  test(`The input will validate when blurred`, () => {
    const wrapper = mountWithTheme(<ExampleTextInput />);
    const input = wrapper.find('input');

    input.simulate('focus');
    input.instance().value = 'invalidemail@';
    input.simulate('change');
    input.simulate('blur');

    expect(
      wrapper.find("[data-test-id='TextInputErrorMessage']")
    ).toMatchSnapshot();
  });
  test(`The input will hide errors at the point it becomes valid`, () => {
    const wrapper = mountWithTheme(<ExampleTextInput />);
    const input = wrapper.find('input');

    input.simulate('focus');
    input.instance().value = 'invalidemail@';
    input.simulate('change');
    input.simulate('blur');

    expect(
      wrapper.find("[data-test-id='TextInputErrorMessage']")
    ).toMatchSnapshot();

    input.simulate('focus');
    input.instance().value = 'valid@email.com';
    input.simulate('change');

    expect(
      wrapper.find("[data-test-id='TextInputErrorMessage']")
    ).toMatchSnapshot();
  });
  test(`A valid input will hide its valid state at the point valid input is added`, () => {
    const wrapper = mountWithTheme(<ExampleTextInput />);
    const input = wrapper.find('input');

    input.simulate('focus');
    input.instance().value = 'valid@email.com';
    input.simulate('change');
    input.simulate('blur');

    expect(
      wrapper.find("[data-test-id='TextInputCheckmark']")
    ).toMatchSnapshot();

    input.simulate('focus');
    input.instance().value = 'valid@email.co';
    input.simulate('change');

    expect(
      wrapper.find("[data-test-id='TextInputCheckmark']")
    ).toMatchSnapshot();
  });

  test(`A valid input will hide its valid state at the point invalid input is added`, () => {
    const wrapper = mountWithTheme(<ExampleTextInput />);
    const input = wrapper.find('input');

    input.simulate('focus');
    input.instance().value = 'valid@email.com';
    input.simulate('change');
    input.simulate('blur');

    input.simulate('focus');
    input.instance().value = 'invalidemail@';
    input.simulate('change');

    expect(
      wrapper.find("[data-test-id='TextInputErrorMessage']")
    ).toMatchSnapshot();
  });
});
