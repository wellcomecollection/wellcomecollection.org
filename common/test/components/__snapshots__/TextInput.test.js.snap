// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput A valid input will hide its valid state at the point invalid input is added 1`] = `null`;

exports[`TextInput A valid input will hide its valid state at the point valid input is added 1`] = `
<span
  className="TextInput__TextInputCheckmark-sc-1860hhu-3 bEVYrx absolute"
  data-test-id="TextInputCheckmark"
>
  <Icon
    extraClasses="icon--green"
    name="check"
  >
    <div
      className="icon icon--green"
    >
      <svg
        aria-hidden={true}
        className="icon__svg"
      >
        <svg
          viewBox="0 0 24 24"
        >
          <path
            className="icon__shape"
            d="M21.13 6.64a1 1 0 0 0-1.41 0L10 16.34l-5-5a1 1 0 0 0-1.41 1.41l5.67 5.67a1 1 0 0 0 1.41 0l10.4-10.4a1 1 0 0 0 .06-1.38z"
            fillRule="nonzero"
          />
        </svg>
      </svg>
    </div>
  </Icon>
</span>
`;

exports[`TextInput A valid input will hide its valid state at the point valid input is added 2`] = `null`;

exports[`TextInput The input will hide errors at the point it becomes valid 1`] = `
<span
  className="TextInput__TextInputErrorMessage-sc-1860hhu-4 IKSWN font-hnm"
  data-test-id="TextInputErrorMessage"
>
  test error message
</span>
`;

exports[`TextInput The input will hide errors at the point it becomes valid 2`] = `null`;

exports[`TextInput The input will validate when blurred 1`] = `
<span
  className="TextInput__TextInputErrorMessage-sc-1860hhu-4 IKSWN font-hnm"
  data-test-id="TextInputErrorMessage"
>
  test error message
</span>
`;
